File: examples/javascript/ttl-caching/simple_websocket.js
// Simple WebSocket Example - Following YCSB benchmark patterns
const WebSocket = require('ws');
const fetch = require('node-fetch');
require('dotenv').config({ path: require('path').join(__dirname, '..', '..', '.env') });

async function simpleWebSocketExample() {
    const BASE_URL = process.env.API_BASE_URL || 'http://localhost:8080';
    const WS_URL = process.env.WS_BASE_URL || 'ws://localhost:8080';
    const API_KEY = process.env.API_BASE_KEY || 'a-test-api-key-from-ekodb';
    
    // Step 1: Get authentication token via HTTP
    let TOKEN = process.env.AUTH_TOKEN;
    if (!TOKEN) {
        const tokenResponse = await fetch(`${BASE_URL}/api/auth/token`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ api_key: API_KEY })
        });
        const tokenData = await tokenResponse.json();
        TOKEN = tokenData.token;
    }

    console.log('✓ Authentication successful');

    try {
        // Step 2: Insert some test data first
        console.log('\n=== Inserting Test Data ===');
        const insertResponse = await fetch(`${BASE_URL}/api/insert/websocket_test`, {
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${TOKEN}`,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                name: 'WebSocket Test Record',
                value: 42,
                active: true
            })
        });
        const insertResult = await insertResponse.json();
        console.log(`✓ Inserted test record: ${insertResult.id}`);

        // Step 3: Connect to WebSocket with Authorization header
        console.log('\n=== Connecting to WebSocket ===');
        const ws = new WebSocket(`${WS_URL}/api/ws`, {
            headers: {
                'Authorization': `Bearer ${TOKEN}`
            }
        });

        // Wait for connection
        await new Promise((resolve, reject) => {
            ws.on('open', () => {
                console.log('✓ WebSocket connected');
                resolve();
            });
            ws.on('error', (error) => {
                console.error('WebSocket error:', error.message);
                reject(error);
            });
        });

        // Step 4: Send a query via WebSocket
        console.log('\n=== Querying Data via WebSocket ===');
        const messageId = String(Date.now());
        
        ws.send(JSON.stringify({
            type: 'FindAll',
            messageId: messageId,
            payload: {
                collection: 'websocket_test'
            }
        }));

        // Wait for response
        const response = await new Promise((resolve, reject) => {
            const timeout = setTimeout(() => {
                reject(new Error('Response timeout'));
            }, 5000);

            ws.on('message', (data) => {
                clearTimeout(timeout);
                const message = JSON.parse(data.toString());
                resolve(message);
            });
        });

        console.log('Response:', JSON.stringify(response, null, 2));
        
        // Show the data we got back
        if (response.payload && response.payload.data) {
            console.log(`✓ Retrieved ${response.payload.data.length} record(s) via WebSocket`);
        }

        // Close connection
        ws.close();
        console.log('\n✓ WebSocket example completed successfully');

    } catch (error) {
        console.error('Error:', error);
        process.exit(1);
    }
}

// Run the example
simpleWebSocketExample().catch(console.error);

File: examples/python/ttl-caching/simple_websocket.py
#!/usr/bin/env python3
"""Simple WebSocket Example - Following YCSB benchmark patterns"""

import asyncio
import os
import json
from pathlib import Path
import aiohttp
import websockets
from dotenv import load_dotenv

# Load environment variables
env_path = Path(__file__).parent.parent.parent / ".env"
load_dotenv(env_path)

BASE_URL = os.getenv("API_BASE_URL", "http://localhost:8080")
WS_URL = os.getenv("WS_BASE_URL", "ws://localhost:8080")
API_KEY = os.getenv("API_BASE_KEY", "a-test-api-key-from-ekodb")


async def get_auth_token():
    """Get authentication token via HTTP"""
    async with aiohttp.ClientSession() as session:
        async with session.post(
            f"{BASE_URL}/api/auth/token",
            json={"api_key": API_KEY},
            headers={"Content-Type": "application/json"},
        ) as response:
            data = await response.json()
            return data["token"]


async def simple_websocket_example():
    """Run simple WebSocket example"""
    # Step 1: Get authentication token
    token = await get_auth_token()
    print("✓ Authentication successful")

    try:
        # Step 2: Insert some test data first
        print("\n=== Inserting Test Data ===")
        async with aiohttp.ClientSession() as session:
            async with session.post(
                f"{BASE_URL}/api/insert/websocket_test",
                headers={
                    "Authorization": f"Bearer {token}",
                    "Content-Type": "application/json",
                },
                json={"name": "WebSocket Test Record", "value": 42, "active": True},
            ) as response:
                result = await response.json()
                print(f'✓ Inserted test record: {result["id"]}')

        # Step 3: Connect to WebSocket with Authorization header
        print("\n=== Connecting to WebSocket ===")
        async with websockets.connect(
            f"{WS_URL}/api/ws", extra_headers={"Authorization": f"Bearer {token}"}
        ) as ws:
            print("✓ WebSocket connected")

            # Step 4: Query data via WebSocket
            print("\n=== Querying Data via WebSocket ===")
            message_id = str(asyncio.get_event_loop().time())

            await ws.send(
                json.dumps(
                    {
                        "type": "FindAll",
                        "messageId": message_id,
                        "payload": {"collection": "websocket_test"},
                    }
                )
            )

            # Wait for response
            response_data = await asyncio.wait_for(ws.recv(), timeout=5.0)
            response = json.loads(response_data)
            print(f"Response: {json.dumps(response, indent=2)}")

            # Show the data we got back
            if response.get("payload") and response["payload"].get("data"):
                print(
                    f'✓ Retrieved {len(response["payload"]["data"])} record(s) via WebSocket'
                )

            print("\n✓ WebSocket example completed successfully")

    except Exception as e:
        print(f"Error: {e}")
        raise


if __name__ == "__main__":
    asyncio.run(simple_websocket_example())
